### DISCUSSION TITLE
Help: ESP32-Audio-Kit A1s LED and  IR receiver

### DISCUSSION THREAD

**Discussion by holefrog** [sle118/squeezelite-esp32#330] [Q&A] [answered]
Hi,
I have an ESP32-Audio-Kit A1s with firmware version I2S-4MFlash-1607, running in Squeezelite-ESP32 mode. 
I want to add a LED and an IR receiver to it, but I have no idea which pins I should use.

**Comment by Renber78** [comment, marked as answer]
You can use any available IO.
For example: If you are not using the SPI port you have a P2 connector that you can use. (GPIO5/18/19/23)
Personally, for the IR receiver, I use GPIO0 in common with BOOT mode. (It is present on connector P3)
Note that on your card you can use one of the two configurable LEDs on the card (GPIO19/22 on P2)
You must then adapt your NVS configuration file (See Readme)

**Reply by holefrog** [reply]
Hi,

I have a VS1838B IR receiver.

I tried to use your settings, "0=ir,  21=amp, 22=green:0, 39=jack:0"
But I have a problem, it will not start. It waits for boot command.
 UART output is:

`picocom  v3.1
port is        : /dev/ttyUSB0
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : no
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        : 
omap is        : 
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no
Type [C-a] [C-h] to see available commands
Terminal ready`

Then I tried to use IO19, "19=ir,  21=amp, 22=green:0, 39=jack:0"
This time, it works, but it cannot decode, UART output:
`I (38848) IR: can't decode IR signal of len 18
I (38848) IR: 10 26 cd 8e 08 29 ef 83 f4 04 c6 83 02 05 b8 83 
I (38848) IR: 14 05 a5 83 ef 04 cb 83 04 05 b6 83 e7 04 ba 83 
I (38868) IR: f6 04 c4 83 20 0e bb 83 90 12 a2 83 91 09 c5 83 
I (38868) IR: 0c 05 98 83 a9 09 ab 83 e7 04 d2 83 1e 05 85 83 
I (38878) IR: 8e 09 c8 83 08 05 00 80 `

 I tried to use IO19, "19=ir:nec,  21=amp, 22=green:0, 39=jack:0", same result.

 I tried to use IO19, "19=ir:rc5,  21=amp, 22=green:0, 39=jack:0", UART  has no output, but LED D5 flashes when I press any key in the remote.

**Reply by holefrog** [reply]
Here is the datasheet of VS1838B:
Features:

    Support 3-5V input voltage,
    The infrared frequency should be 38kHz 1838 NEC code,
    It can sustain a reception distance that goes up to 18 meters,
    High-level voltage: 4.5V,
    Low-level voltage: 0.4V,
    Length at 5.20 mm,
    Width at 0.28inches,
    Height at 7.86 mm,
    Weight at 0.02 oz, and
    Reception angle: 35 degrees.

Specifications:

    Operating voltage: 2.7V to 5.5V,
    Efficient transmission code: RC5 code, NEC code,
    Lead from: 2.54 mm,
    Reception angle: 35-45 degrees,
    Should be CMOS and TTL compatible,
    The  shielding should be of high quality to prevent electric field disturbance, and finally,
    Strong against ambient light.

**Reply by holefrog** [reply]
It is strange, for the same key pressed, each time I have different code.
Like Power button, here is the UART output:
`I (22238) IR: can't decode IR signal of len 18
I (22238) IR: 42 26 b6 8e 14 29 b4 83 26 05 ab 83 f4 04 c6 83 
I (22248) IR: f1 04 c8 83 ff 04 a2 83 2b 05 8f 83 25 05 93 83 
I (22258) IR: 1d 05 b5 83 f0 04 cb 83 94 09 aa 83 a0 12 a9 83 
I (22258) IR: 0e 0e 9f 83 a2 09 b6 83 ee 04 cb 83 0f 05 aa 83 
I (22268) IR: 76 09 c9 83 fb 04 00 80 
I (24068) IR: can't decode IR signal of len 18
I (24068) IR: 16 26 c4 8e 2c 29 cb 83 f3 04 c7 83 ee 04 cb 83 
I (24068) IR: f3 04 c7 83 e3 04 d5 83 e4 04 d6 83 e7 04 b9 83 
I (24088) IR: 0d 05 c5 83 df 04 db 83 71 09 b4 83 a1 12 c2 83 
I (24088) IR: f5 0d d0 83 6a 09 d4 83 06 05 b4 83 f0 04 ca 83 
I (24098) IR: 73 09 ca 83 ea 04 00 80 
I (24748) IR: can't decode IR signal of len 18
I (24748) IR: 39 26 a1 8e 37 29 a8 83 25 05 94 83 1a 05 9f 83 
I (24758) IR: 0c 05 c7 83 12 05 a8 83 e8 04 d1 83 d9 04 e0 83 
I (24768) IR: eb 04 cf 83 0d 05 ac 83 7c 09 c3 83 7e 12 b2 83 
I (24768) IR: 0a 0e ba 83 9f 09 b8 83 ec 04 cd 83 f0 04 ca 83 
I (24778) IR: 78 09 c6 83 f0 04 00 80 
I (37088) IR: can't decode IR signal of len 18
I (37088) IR: 19 26 c6 8e 32 29 ad 83 03 05 b8 83 03 05 cf 83 
I (37088) IR: df 04 c2 83 06 05 b3 83 00 05 d2 83 f1 04 c8 83 
I (37098) IR: e8 04 d2 83 e2 04 d7 83 6e 09 d1 83 70 12 db 83 
I (37098) IR: f0 0d d6 83 73 09 cc 83 e3 04 d8 83 f4 04 ac 83 
I (37108) IR: 8e 09 b1 83 f6 04 00 80 
I (39468) IR: can't decode IR signal of len 18
I (39468) IR: f8 25 e4 8e 2b 29 cd 83 d4 04 e5 83 e2 04 d8 83 
I (39468) IR: ea 04 cf 83 e4 04 d6 83 e5 04 d4 83 e8 04 ba 83 
I (39488) IR: d1 04 00 84 da 04 df 83 6b 09 d4 83 7c 12 cd 83 
I (39488) IR: f9 0d cb 83 63 09 dc 83 a1 04 19 84 ec 04 cd 83 
I (39498) IR: 5d 09 e1 83 e8 04 00 80 
`

**Reply by Renber78** [reply]
Obviously your remote control is not NEC compatible, what are you using?

Recommendations:
- The GPIO19 has a D5 LED, the IR receiver cannot control this LED directly because Icc < 1.5mA!, use GPIO18 instead
- Some old IR sensors work correctly at 5V only
- If your card does not start with GPIO0 you can check that your IR signal is indeed high when it does not receive anything. 
  (This pin is used by the system to enter bootloader mode, it must be high to boot normally)
- The logs you provide are a debug if the code is not recognized (Bad protocol), it is normal that it is not identical on each press. If the remote is compatible, you should only have one line with a hex code.

**Reply by holefrog** [reply]
Hi,

You are right.

- IR signal pin is high when idling. And also the GPIO0 is high when booting.
But it always runs into boot command mode, instead of normal mode, if I plug the IR signal to GPIO0. I do not know why.


- I moved the receiver to GPIO18, now it can decode the command. 
But the problem is the player does not response to the command

Power button
`[00:06:02.904] sendIR:94 sending IR code ffa25d
I (364468) IR: Scan Code (repeat) --- addr: 0x00ff cmd: 0xa25d
`
Play/Pause button
`[00:07:48.360] sendIR:94 sending IR code ffa25d
I (469928) IR: Scan Code (repeat) --- addr: 0x00ff cmd: 0xa25d
`

Volume - button
`[00:08:16.615] sendIR:94 sending IR code ffa857
I (498178) IR: Scan Code (repeat) --- addr: 0x00ff cmd: 0xa857`

**Reply by holefrog** [reply]
Those are the [remote controls ](https://drive.google.com/file/d/1zJI-vXOJqC_sHyWXsrX6LQCqmGunVt2S/view?usp=sharing) I have. They have identical command code.

**Reply by Renber78** [reply]
Well, you're close to the goal.
You must now create your mapping file as explained in this post: https://github.com/sle118/squeezelite-esp32/discussions/306

Below, a config that I made for a remote control similar to yours. It should work and you will probably need to edit it.
[china.zip](https://github.com/sle118/squeezelite-esp32/files/12897996/china.zip)

**Reply by holefrog** [reply]
Thank so much.

Your config and mine almost has no common code, except the power button. I will upload mine when I have it tested.

I have three more problems/questions:
1. I have one Play/Pause button, so I have to use another button (Mode) as 'Play,' and the Play/Pause button as 'Pause.'

2. I cannot find which key is for 'previous' or 'next' in 'Slim_Devices_Remote.ir'. I tried 'previous' and 'next,' as well as 'arrow_left' and 'arrow_right,' but none of them work.

3. The number buttons do not work; for example, pressing 1 does not play the number 1 song.

**Reply by Renber78** [reply]
1 - There is no other solution (at least I think)

2 and 3 - You must have more than one song loaded in your playlist

PS: Thanks for sharing in the remote control post but can you rename your file to "china_remote.ir" because "Slim_Devices_remote.ir" is already used for another existing command.

**Reply by holefrog** [reply]
For 2/3.  Which key should I use, arrow_left or previous?
I am playing a album, and it contains songs.  But one of the key works.
![Screenshot from 2023-10-13 22-15-36](https://github.com/sle118/squeezelite-esp32/assets/17289895/4d810622-6c4e-42ff-aced-ead918827f33)



And for "Slim_Devices_remote.ir", How do I specify which command set LMS uses? I couldn't find the option, I can only replace 'Slim_Devices_remote.ir'.

**Reply by Renber78** [reply]
The commands are "fwd" and "rew" to change songs
The remote control is selectable in: player option / advanced option / remote control

**Reply by holefrog** [reply]
Now the previous/next key  works.  Thanks.
As for "player option / advanced option / remote control", I know the reason now. I have not installed the plugin at that time, so, there is not "remote control" available.
